plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -391,4 +386,97 @@ public int read(byte[] b, int off, int len) throws IOException {
                 "PUT_BLOCK", putBlockOperations.get());
         }
     }
+
+    /**
+     * Building our own input stream instead of using the SDK's {@link com.microsoft.azure.storage.blob.BlobInputStream}
+     * because that stream is highly inefficient in both memory and CPU use.
+     */
+    private static class BlobInputStream extends InputStream {
+
+        /**
+         * Maximum number of bytes to fetch per read request and thus to buffer on heap at a time.
+         * Set to 4M because that's what {@link com.microsoft.azure.storage.blob.BlobInputStream} uses.
+         */
+        private static final int MAX_READ_CHUNK_SIZE = ByteSizeUnit.MB.toIntBytes(4);
+
+        /**
+         * Using a {@link ByteArrayOutputStream} as a buffer instead of a byte array since the byte array APIs on the SDK are less
+         * efficient.
+         */
+        private final ByteArrayOutputStream buffer;
+
+        private final long limit;
+
+        private final CloudBlockBlob blockBlobReference;
+
+        private final long start;
+
+        private final OperationContext context;
+
+        // current read position on the byte array backing #buffer
+        private int pos;
+
+        // current position up to which the contents of the blob where buffered
+        private long offset;
+
+        BlobInputStream(long limit, CloudBlockBlob blockBlobReference, long start, OperationContext context) {
+            this.limit = limit;
+            this.blockBlobReference = blockBlobReference;
+            this.start = start;
+            this.context = context;
+            buffer = new ByteArrayOutputStream(Math.min(MAX_READ_CHUNK_SIZE, Math.toIntExact(Math.min(limit, Integer.MAX_VALUE)))) {
```

Comment:
Still horrific to allocate `4M` on heap but unlike the SDKs stream we at least reuse these bytes instead of allocating them over and over for larger blobs.

Reply:
Maybe we can write our own `ByteArrayOutputStream` version as all methods are synchronized and `buffer.toByteArray()` returns a copy that we [copy](https://github.com/elastic/elasticsearch/pull/61812/files#diff-1e589690e786a5607132330247f7da34R463) again?.

Reply:
I do override `toByteArray` below to prevent that? :)

As far as the `synchronized` goes, I'm not sure it's worth the extra code. I would expect JIT to take care of that since we're never concurrently accessing the BAOS. Probably not worth the extra code? (also when I profiled this with @hub-cap the copying into `buffer` took trivial time only so not much to gain anyway)

Reply:
Right, I missed that 🤦. Thanks for clarifying :)